jwst/pipeline/calwebb_tso3.py

```diff
@@ -0,0 +1,143 @@
+        """
+        Run the TSO3Pipeline
+
+        Parameters
+        ----------
+        input: Level3 Association, or ModelContainer
+            The exposures to process
+        """
+
+        self.log.info('Starting calwebb_tso3 ...')
+        print("scale_detection set to : {}".format(self.scale_detection))
+        
+        input_asn = datamodels.open(input)
+        product_name = input_asn.meta.asn_table.products[0].name
+
+        # Setup output creation
+        #self.output_basename = product['name']
+        self.output_basename = product_name 
+        
+        # Input may consist of multiple exposures, so loop over each of them
+        for cube in input_asn:
+            # Convert CubeModel into ModelContainer of 2-D DataModels
+            input_models = datamodels.ModelContainer()
+            for i in range(cube.data.shape[0]):
+                # convert each plane of data cube into it own array
+                # for outlier detection...
+                image = datamodels.ImageModel(data=cube.data[i],
+                        err=cube.err[i], dq=cube.dq[i])
+                image.meta = cube.meta
+                input_models.append(image)
+            
+            if not self.scale_detection:
+                self.log.info("Performing outlier detection on input images...")
+                results = self.outlier_detection(input_models)
+            else:
+                self.log.info("Performing scaled outlier detection on input images...")
+                results = self.outlier_detection_scaled(input_models)
+                
+            # Transfer updated DQ values to original input observation
+            for i in range(cube.data.shape[0]):
+                # preserve output filename 
+                orig_filename = cube.meta.filename
+                # Update DQ arrays with those from outlier_detection step
+                cube.dq[i] = input_models[i].dq
+                # reset output filename to original value
+                cube.meta.filename = orig_filename
+
+        self.log.info("Writing Level 2c images with updated DQ arrays...")
+        suffix_2c = 'crfints'
+        for cube in input_asn:
+            self.output_basename = cube.meta.filename
+            self.save_model(cube, suffix=suffix_2c)
+
+        # Create final photometry results as a single output
+        # regardless of how many members their may be...
+        phot_result_list = []
+        if 'image' in results[0].meta.exposure.type.lower():
+            # Create name for extracted photometry (Level 3) products
+            phot_tab_name = "{}_phot.ecsv".format(product_name)
+
+            # For each cube, extract photometry...
+            for cube in input_asn:
+                # Extract Photometry from imaging data
+                phot_result_list.append(self.tso_photometry(cube))
+        else:
+            # Working with spectroscopic TSO data...
+            # define output for x1d (level 2c/3) products
```

Comment:
Technically all output products from here on are only level 3. The 2c's were already created above.